Validation/MuonHits/src/CSCSimHitMatcher.cc

```diff
@@ -0,0 +1,387 @@
+      if (MuonHitHelper::toCSCType(detId.station(), detId.ring()) != csc_type)
+        continue;
+    }
+    result.insert(id);
+  }
+  return result;
+}
+
+int CSCSimHitMatcher::nLayersWithHitsInChamber(unsigned int detid) const {
+  set<int> layers_with_hits;
+  for (const auto& h : MuonSimHitMatcher::hitsInChamber(detid)) {
+    const CSCDetId& idd(h.detUnitId());
+    layers_with_hits.insert(idd.layer());
+  }
+  return layers_with_hits.size();
+}
+
+bool CSCSimHitMatcher::hitStation(int st, int nlayers) const {
+  int nst = 0;
+  for (const auto& ddt : chamberIds()) {
+    const CSCDetId id(ddt);
+    int ri(id.ring());
+    if (id.station() != st) continue;
+
+    // ME1/a case - check the ME1/b chamber
+    if (st == 1 and ri == 4) {
+      CSCDetId idME1b(id.endcap(), id.station(), 1, id.chamber(), 0);
+      const int nl1a(nLayersWithHitsInChamber(id.rawId()));
+      const int nl1b(nLayersWithHitsInChamber(idME1b.rawId()));
+      if (nl1a + nl1b < nlayers) continue;
+      ++nst;
+    }
+    // ME1/b case - check the ME1/a chamber
+    else if (st == 1 and ri == 1) {
+      CSCDetId idME1a(id.endcap(), id.station(), 4, id.chamber(), 0);
+      const int nl1a(nLayersWithHitsInChamber(idME1a.rawId()));
+      const int nl1b(nLayersWithHitsInChamber(id.rawId()));
+      if (nl1a + nl1b < nlayers) continue;
+      ++nst;
+    }
+    // default case
+    else {
+      const int nl(nLayersWithHitsInChamber(id.rawId()));
+      if (nl < nlayers) continue;
+      ++nst;
+    }
+  }
+  return nst;
+}
+
+int CSCSimHitMatcher::nStations(int nlayers) const {
+  return (hitStation(1, nlayers) + hitStation(2, nlayers) +
+          hitStation(3, nlayers) + hitStation(4, nlayers));
+}
+
+float CSCSimHitMatcher::LocalBendingInChamber(unsigned int detid) const {
+  const CSCDetId cscid(detid);
+  if (nLayersWithHitsInChamber(detid) < 6) return -100;
+  float phi_layer1 = -10;
+  float phi_layer6 = 10;
+
+  if (cscid.station() == 1 and (cscid.ring() == 1 or cscid.ring() == 4)) {
+    // phi in layer 1
+    const CSCDetId cscid1a(cscid.endcap(), cscid.station(), 4, cscid.chamber(),
+                           1);
+    const CSCDetId cscid1b(cscid.endcap(), cscid.station(), 1, cscid.chamber(),
+                           1);
+    const edm::PSimHitContainer& hits1a =
+        MuonSimHitMatcher::hitsInDetId(cscid1a.rawId());
+    const edm::PSimHitContainer& hits1b =
+        MuonSimHitMatcher::hitsInDetId(cscid1b.rawId());
+    const GlobalPoint& gp1a =
+        simHitsMeanPosition(MuonSimHitMatcher::hitsInDetId(cscid1a.rawId()));
+    const GlobalPoint& gp1b =
+        simHitsMeanPosition(MuonSimHitMatcher::hitsInDetId(cscid1b.rawId()));
+    if (!hits1a.empty() and !hits1b.empty())
+      phi_layer1 = (gp1a.phi() + gp1b.phi()) / 2.0;
+    else if (!hits1a.empty())
+      phi_layer1 = gp1a.phi();
+    else if (!hits1b.empty())
+      phi_layer1 = gp1b.phi();
+    else
+      std::cerr << " no hits in layer1, cant not find global phi of hits "
```

Comment:
@dildick all the cerr instances in this method have the same non thread-safety issues of cout, should be replaced by LogError (or LogWarning or whatever else, depending on the severity you want to attribute to the problem)